drivers/source/UnbufferedSerial.cpp

```diff
@@ -0,0 +1,111 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2019 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "drivers/UnbufferedSerial.h"
+
+#if DEVICE_SERIAL
+
+#include "platform/mbed_critical.h"
+
+namespace mbed {
+
+UnbufferedSerial::UnbufferedSerial(
+    PinName tx,
+    PinName rx,
+    int baud
+) : SerialBase(tx, rx, baud)
+{
+    // No lock needed in the constructor
+}
+
+ssize_t UnbufferedSerial::write(const void *buffer, size_t size)
+{
+    const unsigned char *buf = static_cast<const unsigned char *>(buffer);
+
+    if (size == 0) {
+        return 0;
+    }
+
+    bool lock_api = !core_util_in_critical_section();
+
+    if (lock_api) {
+        lock();
+    }
+
+    for (size_t i = 0; i < size; i++) {
+        _base_putc(buf[i]);
+    }
+
+    if (lock_api) {
+        unlock();
+    }
+
+    return size;
+}
+
+ssize_t UnbufferedSerial::read(void *buffer, size_t size)
+{
+    unsigned char *buf = static_cast<unsigned char *>(buffer);
+
+    if (size == 0) {
+        return 0;
+    }
+
+    bool lock_api = !core_util_in_critical_section();
```

Comment:
Don't think it's worth having this on `read`. Only reason to have it on `write` is for emergency crash output.

Reply:
I did not have it because I thought it was not necessary but then you wrote [this](https://github.com/ARMmbed/mbed-os/pull/11961#discussion_r352560931) and I thought you meant I should put it because I also have it in `write` or did you mean something else with that comment?

Reply:
You should have the lock. I'm saying you don't need the `core_util_in_critical_section` emergency lock bypass.

(It's even a bit overkill for the `write` - only needed if someone overrides your `lock`, but I guess that's possible).

Reply:
[Done](https://github.com/ARMmbed/mbed-os/compare/ed45befc9a6f933a9c2a5b4d6781874d9013df79..a39e03b62718502832426ed4b463e4878bb7d65b).